sklearn/callback/_progressbar.py

```diff
@@ -0,0 +1,187 @@
+# License: BSD 3 clause
+# Authors: the scikit-learn developers
+
+import importlib
+from multiprocessing import Manager
+from threading import Thread
+
+from . import BaseCallback
+
+
+class ProgressBar(BaseCallback):
+    """Callback that displays progress bars for each iterative steps of an estimator"""
+
+    auto_propagate = True
+
+    def __init__(self):
+        try:
+            importlib.import_module("rich")  # noqa
+        except ImportError as e:
+            raise ImportError("ProgressBar requires rich installed.") from e
+
+    def on_fit_begin(self, estimator, X=None, y=None):
+        self._queue = Manager().Queue()
+        self.progress_monitor = _RichProgressMonitor(queue=self._queue)
+        self.progress_monitor.start()
+
+    def on_fit_iter_end(self, *, estimator, node, **kwargs):
+        self._queue.put(node)
+
+    def on_fit_end(self):
+        self._queue.put(None)
+        self.progress_monitor.join()
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        if "progress_monitor" in state:
+            del state["progress_monitor"]
+        return state
+
+
+# Custom Progress class to allow showing the tasks in a given order (given by setting
+# the _ordered_tasks attribute). In particular it allows to dynamically create and
+# insert tasks between existing tasks.
+
+try:
+    from rich.progress import BarColumn, Progress, TextColumn, TimeRemainingColumn
+    from rich.style import Style
+
+    class _Progress(Progress):
+        def get_renderables(self):
+            table = self.make_tasks_table(getattr(self, "_ordered_tasks", []))
+            yield table
+
+except ImportError:
+    pass
+
+
+class _RichProgressMonitor(Thread):
+    """Thread monitoring the progress of an estimator with rich based display
+
+    The display is a list of nested rich tasks using rich.Progress. There is one for
+    each non-leaf node in the computation tree of the estimator.
+
+    Parameters
+    ----------
+    queue : multiprocessing.Manager.Queue instance
+        This thread will run until the queue is empty.
+    """
+
+    def __init__(self, *, queue):
+        Thread.__init__(self)
+        self.queue = queue
+
+    def run(self):
+        self.progress_ctx = _Progress(
+            TextColumn("[progress.description]{task.description}"),
+            BarColumn(
+                complete_style=Style(color="dark_orange"),
+                finished_style=Style(color="cyan"),
```

Comment:
As a future feature, we could think on allowing a color_config to be passed to the callable to be customable. But I would rather don't bother for the moment.